tests/cpptests/test_cpp_iterator_union.cpp

```diff
@@ -0,0 +1,268 @@
+/*
+ * Copyright Redis Ltd. 2016 - present
+ * Licensed under your choice of the Redis Source Available License 2.0 (RSALv2) or
+ * the Server Side Public License v1 (SSPLv1).
+ */
+
+#include "rmutil/alloc.h"
+
+#include "gtest/gtest.h"
+#include "iterator_util.h"
+
+#include "src/iterators/union_iterator.h"
+
+class UnionIteratorCommonTest : public ::testing::TestWithParam<std::tuple<unsigned, bool, std::vector<t_docId>>> {
+protected:
+    std::vector<std::vector<t_docId>> docIds;
+    std::vector<t_docId> resultSet;
+    QueryIterator *ui_base;
+
+    void SetUp() override {
+        ASSERT_EQ(RSGlobalConfig.iteratorsConfigParams.minUnionIterHeap, 20) <<
+            "If we ever change the default threshold for using heaps, we need to modify the tests "
+            "here so they still check both flat and heap alternatives.";
+
+        auto [numChildren, quickExit, union_res] = GetParam();
+        // Set resultSet to the expected union result
+        resultSet = union_res;
+        // Set docIds so the union of all children is union_res.
+        // Make sure that some ids are repeated in some children
+        docIds.resize(numChildren);
+        for (size_t i = 0; i < union_res.size(); i++) {
+            for (unsigned j = 0; j < numChildren; j++) {
+                if (j % (i + 1) == 0) {
+                    docIds[j].push_back(union_res[i]);
+                }
+            }
+        }
+        // Create children iterators
+        auto children = (QueryIterator **)rm_malloc(sizeof(QueryIterator *) * numChildren);
+        for (unsigned i = 0; i < numChildren; i++) {
+            children[i] = (QueryIterator *) new MockIterator(docIds[i]);
+        }
+        // Create a union iterator
+        ui_base = IT_V2(NewUnionIterator)(children, numChildren, quickExit, 1.0, QN_UNION, NULL, &RSGlobalConfig.iteratorsConfigParams);
+    }
+    void TearDown() override {
+        ui_base->Free(ui_base);
+    }
+};
+
+TEST_P(UnionIteratorCommonTest, Read) {
+    UnionIterator *ui = (UnionIterator *)ui_base;
+    IteratorStatus rc;
+
+    // Test reading until EOF
+    size_t i = 0;
+    while ((rc = ui_base->Read(ui_base)) == ITERATOR_OK) {
+        ASSERT_EQ(ui->base.current->docId, resultSet[i]);
+        ASSERT_EQ(ui->base.lastDocId, resultSet[i]);
+        ASSERT_TRUE(ui->base.isValid);
+        i++;
+    }
+    ASSERT_EQ(rc, ITERATOR_EOF);
+    ASSERT_FALSE(ui->base.isValid);
+    ASSERT_EQ(ui_base->Read(ui_base), ITERATOR_EOF); // Reading after EOF should return EOF
+    ASSERT_EQ(i, resultSet.size()) << "Expected to read " << resultSet.size() << " documents";
+
+    size_t expected = 0;
+    for (auto &child : docIds) {
+        expected += child.size();
+    }
+    ASSERT_EQ(ui->num_results_estimated, expected);
+    ASSERT_EQ(ui_base->NumEstimated(ui_base), expected);
+}
+
+TEST_P(UnionIteratorCommonTest, SkipTo) {
+    UnionIterator *ui = (UnionIterator *)ui_base;
```

Comment:
the most of the test is using `ui_base` and not `ui`. aimed?

Reply:
Most of the time I don't have anything to check that I cannot access from the base struct (API, current, lastDocId). Didn't want to use `ui->base.`, `&ui->base` if I don't have to